From f3f0f27f1463116b6d3945aad9f9f8ce98d3a15b Mon Sep 17 00:00:00 2001
From: Elena Reshetova <elena.reshetova@intel.com>
Date: Tue, 14 Jun 2022 11:59:52 +0300
Subject: [PATCH 1405/1418] efi/x86-stub: fix absolute symbol reference

Fixes the following build error on ubuntu 22.04:

0000000000000000 R_X86_64_64 .text+0x00000000000003b9
0000000000000000 R_X86_64_64 .text+0x00000000000003b9
drivers/firmware/efi/libstub/x86-stub.stub.o: absolute symbol
references not allowed in the EFI stub
make[6]: *** [/home/elena/tdx/tdx-tools/build/ubuntu-22.04/
intel-mvp-spr-kernel/mvp-linux-kernel-5.15/drivers/firmware/
efi/libstub/Makefile:129:
drivers/firmware/efi/libstub/x86-stub.stub.o] Error 1

Signed-off-by: Elena Reshetova <elena.reshetova@intel.com>
---
 drivers/firmware/efi/libstub/x86-stub.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/firmware/efi/libstub/x86-stub.c b/drivers/firmware/efi/libstub/x86-stub.c
index 26f47210cb14..68855d10953c 100644
--- a/drivers/firmware/efi/libstub/x86-stub.c
+++ b/drivers/firmware/efi/libstub/x86-stub.c
@@ -36,7 +36,12 @@ static void detect_intel_tdx(void)
 {
 	u32 eax, sig[3];
 
-	cpuid_count(TDX_CPUID_LEAF_ID, 0, &eax, &sig[0], &sig[2], &sig[1]);
+	asm volatile(".ifnc %%ebx,%3 ; movl  %%ebx,%3 ; .endif	\n\t"
+		     "cpuid					\n\t"
+		     ".ifnc %%ebx,%3 ; xchgl %%ebx,%3 ; .endif	\n\t"
+		    : "=a" (eax), "=c" (sig[2]), "=d" (sig[1]), "=b" (sig[0])
+		    : "a" (TDX_CPUID_LEAF_ID), "c" (0)
+	);
 
 	if (memcmp("IntelTDX    ", sig, 12))
 		return;
-- 
2.31.1

